### PR TITLE
added copy url button

### LINKED ISSUE #8802: Add "Copy URL" Option to Share Modal
Especially on mobile, there is no way to copy the url of the page you're on in order to share it. Our current share modal has an "embed" option, which is similar, but we want to add an option for copy url:

Screenshot 2024-02-07 at 10 23 36 AM

Screenshot 2024-02-07 at 10 23 22 AM

We can use this icon:

https://www.flaticon.com/free-icon/link_341077

image

### Steps to Reproduce

1. Go to ...
2. Do ...

* Actual:
* Expected:

### Details

- **Logged in (Y/N)?**
- **Browser type/version?**
- **Operating system?**
- **Environment (prod/dev/local)?** prod

### PR DESCRIPTION
**Pull request by Ayush1404**
Closes #8802

Adds way to copy the URL of the page you're on to share it.

### Testing

Go to any books page and click on share.
### Screenshot

Screenshot from 2024-02-08 20-28-16

### Stakeholders

@mekarpeles

### CODE REVIEW

**Comment by mekarpeles** [comment]
@jimchamp is this a good place to use toast? Do we have an example of doing this via js?

**Comment by jimchamp** [comment]
> @jimchamp is this a good place to use toast? Do we have an example of doing this via js?

Yes, and the modal should close when the URL copy icon is clicked.  Will provide guidance shortly...

File: openlibrary/macros/ShareModal.html [hunk 1]
```diff
@@ -49,6 +49,20 @@
               </a>
               <div class="share-source">$_('Embed')</div>
             </li>
+
+            <li>
+              <a class="copy-url-btn" title="$_('Copy url to clipboard')" data-ol-link-track="Share|CopyURL"
+                  onclick="navigator.clipboard.writeText('openlibrary.org'+'$page_url');">
+                  <img
+                      alt="CopyURL icon"
```
Comment (jimchamp) [inline openlibrary/macros/ShareModal.html:57]:
```suggestion
                      alt="$_('Copy URL icon')"
```
Add space between `Copy` and `URL`.  Use localized strings (more information about this [here](https://github.com/internetarchive/openlibrary/wiki/Internationalization-Contributor's-Guide-(i18n)#internationalization-i18n-developers-guide)).

File: openlibrary/macros/ShareModal.html [hunk 2]
```diff
@@ -49,6 +49,20 @@
               </a>
               <div class="share-source">$_('Embed')</div>
             </li>
+
+            <li>
+              <a class="copy-url-btn" title="$_('Copy url to clipboard')" data-ol-link-track="Share|CopyURL"
+                  onclick="navigator.clipboard.writeText('openlibrary.org'+'$page_url');">
```
Comment (jimchamp) [inline openlibrary/macros/ShareModal.html:55]:
```suggestion
              <a class="copy-url-btn" title="$_('Copy url to clipboard')" data-ol-link-track="Share|CopyURL">
```
Move inline JS to Webpack.  To do so, you can update [this function](https://github.com/internetarchive/openlibrary/blob/8ca25d920acf98ea431f0fae1e9e7876339315aa/openlibrary/plugins/openlibrary/js/modals/index.js#L5-L7) to do the following:

1. Get a reference to the "Copy URL" icon.  This query should work `#social-modal-content .copy-url-btn`.
2. Add a click listener to the icon.

The click listener should:
1. Copy the URL to the clipboard.
2. Close the modal. (Running `$.colorbox.close()` will achieve this)
3. Display a toast message that tells the patron that the message has been copied to their clipboard.  You can use [this function](https://github.com/internetarchive/openlibrary/blob/8ca25d920acf98ea431f0fae1e9e7876339315aa/openlibrary/plugins/openlibrary/js/modals/index.js#L172) to display the toast message (only pass `message` --- `$parent` is not needed).

**Review by jimchamp — CHANGES_REQUESTED** [review]
Thanks, @Ayush1404!  Happy to merge this once the suggested changes have been made.

You may already be aware of this, but when any `.less` or `.js` file is modified, the CSS and JS assets must be recompiled.  There is more information about this [here](https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide#css-js-and-html).

Let me know if you have any questions about anything.

**Comment by Ayush1404** [comment]
Hey @jimchamp, I have done all the corrections suggested by you. While doing so I realized that embed-work-btn button also has some js in Share_Modal.html [Here](https://github.com/internetarchive/openlibrary/blob/40b17fc3fd4e0ff92549723ef433dfbef5cc548d/openlibrary/macros/ShareModal.html#L42C17-L42C83) . Should I move that code also to Webpack?

**Comment by Ayush1404** [comment]
Hey @jimchamp, I have done all the corrections suggested by you. While doing so I realized that embed-work-btn button also has some js in Share_Modal.html [Here](https://github.com/internetarchive/openlibrary/blob/40b17fc3fd4e0ff92549723ef433dfbef5cc548d/openlibrary/macros/ShareModal.html#L42C17-L42C83) . Should I move that code also to Webpack?

File: openlibrary/plugins/openlibrary/js/modals/index.js [hunk 3]
```diff
@@ -2,8 +2,24 @@ import { FadingToast } from '../Toast.js';
...
+    addShareModalButtonListeners();
+}
+/**
+ * Adds click listeners to buttons in all notes modals on a page.
+ */
+function addShareModalButtonListeners (){
+    $('#social-modal-content .copy-url-btn').on('click', function(event){
+        event.preventDefault();
+        navigator.clipboard.writeText(window.location.href);
+        showToast('URL copied successfully')
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/modals/index.js:20]: @jimchamp question, is there a way we handle i18n in JS like this?

Comment (jimchamp) [reply openlibrary/plugins/openlibrary/js/modals/index.js:20]: Yes.  I'll include that in the follow-up issue (mentioned [here](https://github.com/internetarchive/openlibrary/pull/8806#issuecomment-1971633853)).

File: openlibrary/plugins/openlibrary/js/modals/index.js [hunk 4]
```diff
@@ -2,8 +2,24 @@ import { FadingToast } from '../Toast.js';
 import '../../../../../static/css/components/metadata-form.less';
 
 
+/**
+ * Initializes share modal.
+ */
+
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/modals/index.js:8]:
```suggestion
```

File: openlibrary/plugins/openlibrary/js/modals/index.js [hunk 5]
```diff
@@ -2,8 +2,24 @@ import { FadingToast } from '../Toast.js';
...
 export function initShareModal($modalLinks) {
     addClickListeners($modalLinks, '400px');
+    addShareModalButtonListeners();
+}
+/**
+ * Adds click listeners to buttons in all notes modals on a page.
+ */
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/modals/index.js:15]:
```suggestion

/**
 * Adds click listener which copies the current page's URL to the clipboard.
 */
```

File: openlibrary/plugins/openlibrary/js/modals/index.js (hunk 3 above)
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/modals/index.js:20]:
```suggestion
        showToast('URL copied to clipboard')
```
Update for clarity.

**Review by jimchamp — APPROVED** [review]
Thanks, @Ayush1404!

I'm going to commit my small suggested changes, then merge this branch.

**Comment by jimchamp** [comment]
I ended up moving the inline `embed html` code to Webpack in #8853.

**Pull request merged**